<div *ngIf="reference && comparison" id="browser-container">
  <h2>{{ reference.commonName }}, Chr{{ chromosome }}</h2>

  <div class="btn-group">
    <button class="btn btn-outline" (click)="panLeft()" [disabled]="!interval || interval.start === 0">
      <clr-icon shape="rewind"></clr-icon> Pan Left
    </button>
    <button class="btn btn-outline" (click)="zoomOut()" [disabled]="!interval || interval.width === getRefChrSize()">
      <clr-icon shape="zoom-out"></clr-icon> Zoom Out
    </button>
    <clr-dropdown *ngIf="featuresAreSelected()">
      <button type="button" class="btn btn-outline group-dropdown" clrDropdownTrigger [disabled]="!interval">
        Jump to Feature
        <clr-icon shape="caret down"></clr-icon>
      </button>
      <clr-dropdown-menu clrPosition="bottom-right" *clrIfOpen>
        <div *ngIf="this.selectedFeatures.reference.length > 0">
          <label class="dropdown-header">Genes</label>
          <button *ngFor="let gene of selectedFeatures.reference" type="button" clrDropdownItem (click)="jumpTo(gene)">
            {{ gene.symbol }}
          </button>
        </div>
        <div *ngIf="selectedQTLs.length > 0 && this.selectedFeatures.reference.length > 0" class="dropdown-divider"></div>
        <div *ngIf="selectedQTLs.length > 0">
          <label class="dropdown-header">QTLs</label>
          <button *ngFor="let qtl of selectedQTLs" type="button" clrDropdownItem (click)="jumpTo(qtl)">
            {{ qtl.qtl_symbol }}
          </button>
        </div>

      </clr-dropdown-menu>
    </clr-dropdown>
    <button class="btn btn-outline" (click)="zoomIn()" [disabled]="!interval || interval.width === minimumIntervalSize">
      <clr-icon shape="zoom-in"></clr-icon> Zoom In
    </button>
    <button class="btn btn-outline" (click)="panRight()" [disabled]="!interval || interval.end === getRefChrSize()">
      <clr-icon shape="fast-forward"></clr-icon> Pan Right
    </button>
  </div>

  <svg class="block-view" [attr.viewBox]="'0 0 ' + width + ' ' + height" preserveAspectRatio="xMinYMin meet">
    <!-- Chromosome View -->
    <g id="chromosome-view">
      <!-- Indicators for selected and filtered genes in the reference chromosome -->
      <path *ngFor="let gene of selectedFeatures.reference" class="gene-mark"
            [attr.d]="getVLinePath(gene.getCenterXPos(staticRefBPToPixels), 8, 5)">
      </path>

      <!-- Chromosome view syntenic blocks (and background) -->
      <g id="chr-view-blocks">
        <!-- Syntenic blocks -->
        <rect *ngFor="let block of blocks" [attr.transform]="translate(staticRefBPToPixels(block.ref_start), 15)"
              [attr.width]="staticRefBPToPixels(block.ref_end - block.ref_start)" [attr.height]="chromosomeViewHeight - 25"
              [ngStyle]="{ 'fill': getBlockColor(block), 'stroke': getBlockColor(block) }">
        </rect>

        <!-- Indicators for selected QTLs -->
        <path *ngFor="let qtl of selectedQTLs" class="qtl qtl-mark"
              [attr.d]="getHLinePath(getQTLMarkY(qtl), getQTLWidth(qtl, staticRefBPToPixels), staticRefBPToPixels(qtl.start))">
        </path>

        <!-- Transparent cover for brush (and when rendered, brush elements) -->
        <g id="chr-view-inv-cover"></g>
      </g>

      <!-- Indicators for selected and filtered features in the comparison chromsome -->
      <path *ngFor="let gene of selectedFeatures.comparison" class="gene-mark"
            [attr.d]="getVLinePath(gene.getCenterXPos(getStaticCompScale(gene)), 8, chromosomeViewHeight - 8)">
      </path>

      <!-- Static axis for chromsome view -->
      <g id="chr-view-axis" [attr.transform]="translate(0, chromosomeViewHeight + 5)"></g>
    </g>

    <!-- Horizontal rule to visually separate chromsome view and browser -->
    <path class="hor-rule" [attr.d]="getHLinePath(browserOffset - 50, width)"></path>

    <!-- Synteny browser -->
    <g id="browser" [attr.transform]="translate(0, browserOffset)">
      <!-- Backgrounds -->
      <text class="all-caps" [attr.transform]="translate(2, trackHeight + 16)">{{ reference.commonName }}</text>

      <rect class="trans-bg bg" [attr.height]="(2 * trackHeight) + 40"></rect>

      <!-- Dynamic axis that scales with browser scaling for more location context of reference track -->
      <g id="browser-axis" [attr.transform]="translate(0, -15)"></g>

      <!-- Reference chromsome track -->
      <g id="reference-genome">
        <!-- Reference block coordinates -->
        <g *ngIf="blocks" id="reference-block-coordinates">
          <g *ngFor="let block of blocks" [attr.transform]="translate(refBPToPixels(block.ref_start), 0)">
            <g *ngIf="getBlockWidth(block) > 115">
              <!-- Vertical line extending start point of block -->
              <path [attr.d]="getVLinePath(0, -8)"></path>

              <!-- Starting block coordinate -->
              <text [attr.transform]="translate(2, -2)"> {{ block.ref_chr }}:{{ block.ref_start }} </text>

              <!-- Ending block coordinate -->
              <text class="end" [attr.transform]="translate(getBlockWidth(block) - 2, -2)">
                {{ block.ref_chr }}:{{ block.ref_end }}
              </text>

              <!-- Vertical line extending end point of block -->
              <path [attr.d]="getVLinePath(getBlockWidth(block), -8)"></path>
            </g>
          </g>
        </g>

        <!-- Reference track syntenic blocks -->
        <rect *ngFor="let block of blocks" class="reference" [attr.x]="refBPToPixels(block.ref_start)"
              [attr.width]="getBlockWidth(block)" [attr.height]="trackHeight" >
        </rect>

        <!-- Selected QTLs -->
        <rect *ngFor="let qtl of selectedQTLs" class="qtl" [attr.transform]="translate(refBPToPixels(qtl.start), 0)"
              [attr.width]="getQTLWidth(qtl, refBPToPixels)" [attr.height]="trackHeight">
        </rect>

        <!-- Reference genes -->
        <g *ngIf="referenceGenes">
          <g *ngFor="let gene of getRefGenesInView()" (mouseenter)="highlightRef(gene)" (mouseleave)="unhighlight()"
             [attr.transform]="translate(gene.getXPos(refBPToPixels), gene.getYPos(trackHeight))">
            <!-- Gene label -->
            <text *ngIf="gene.hasVisibleLabel()" [ngStyle]="{ 'fill': gene.getColor() }">
              {{ gene.symbol }}
            </text>

            <!-- Gene -->
            <rect class="gene" [attr.width]="gene.getWidth(refBPToPixels)" [ngStyle]="{ 'fill': gene.getColor() }"></rect>

            <!-- Transcript -->
            <g *ngIf="interval.width <= 5000000">
              <rect *ngFor="let exon of gene.transcript" class="exon" [ngStyle]="{ 'fill': gene.getColor() }"
                    [attr.width]="gene.getExonWidth(exon, refBPToPixels)"
                    [attr.transform]="translate(gene.getRefExonXPos(exon, refBPToPixels), 0)"></rect>
            </g>
          </g>
        </g>
      </g>

      <!-- Orientation indicators -->
      <g *ngIf="blocks && blockOrientation === 'match_orientation'" id="orientation-indicators">
        <path *ngFor="let block of getNonMatchedBlocks()" [attr.d]="getOrientationIndPathCommand(block)"></path>
      </g>

      <!-- Comparison chromosome track -->
      <g id="comparison-genome" [attr.transform]="translate(0, trackHeight + 30)">
        <!-- Comparison track syntenic blocks -->
        <rect *ngFor="let block of blocks" class="comparison" [attr.x]="refBPToPixels(block.ref_start)"
              [attr.width]="getBlockWidth(block)" [attr.height]="trackHeight"
              [ngStyle]="{ 'fill': getBlockColor(block), 'stroke': getBlockColor(block) }">
      </rect>

        <!-- Comparison genes -->
        <g *ngIf="comparisonGenes && compBPToPixels && blocks" id="comparison-genes">
          <g *ngFor="let gene of getCompGenesInView()" (mouseenter)="highlightComp(gene)" (mouseleave)="unhighlight()"
             [attr.transform]="translate(gene.getXPos(getScale(gene), getStart(gene)), gene.getYPos(trackHeight))">
            <!-- Gene label -->
            <text *ngIf="gene.hasVisibleLabel()" [ngStyle]="{ 'fill': gene.getColor() }">
              {{ gene.symbol }}
            </text>

            <!-- Gene -->
            <rect class="gene" [attr.width]="gene.getWidth(getScale(gene))" [ngStyle]="{ 'fill': gene.getColor() }"></rect>

            <!-- Transcript -->
            <g *ngIf="interval.width <= 5000000">
              <rect *ngFor="let exon of gene.transcript" class="exon" [ngStyle]="{ 'fill': gene.getColor() }"
                    [attr.width]="gene.getExonWidth(exon, getScale(gene))"
                    [attr.transform]="translate(gene.getCompExonXPos(exon, getScale(gene), getStart(gene)), 0)"></rect>
            </g>
          </g>
        </g>

        <!-- Comparison block coordinates -->
        <g *ngIf="blocks" id="comparison-block-coordinates">
          <g *ngFor="let block of blocks" [attr.transform]="translate(refBPToPixels(block.ref_start), trackHeight + 9)">
            <g *ngIf="getBlockWidth(block) > 115">
              <!-- Vertical line extending start point of block -->
              <path [attr.d]="getVLinePath(0, -10)" [attr.stroke]="getBlockColor(block)"></path>

              <!-- Starting block coordinate -->
              <text [attr.transform]="translate(2, -1)">
                {{ block.comp_chr }}:{{ block[blockOrientation].comp_start }}
              </text>

              <!-- Ending block coordinate -->
              <text class="end" [attr.transform]="translate(getBlockWidth(block) - 2, -1)">
                {{ block.comp_chr }}:{{ block[blockOrientation].comp_start }}
              </text>

              <!-- Vertical line extending end point of block -->
              <path [attr.d]="getVLinePath(getBlockWidth(block), -10)" [attr.stroke]="getBlockColor(block)"></path>
            </g>
          </g>
        </g>
      </g>

      <!-- Interval coordinates for current view extents in reference positions -->
      <g *ngIf="interval" id="reference-interval-coordinates" [attr.transform]="translate(0, -35)">
        <!-- Vertical line extending start point of reference interval -->
        <path [attr.d]="getVLinePath(0, trackHeight + 43, -8)"></path>

        <!-- Start position -->
        <text [attr.transform]="translate(2, 0)">
          {{ reference.commonName }}, Chr{{ chromosome }}:{{ interval.start }} b
        </text>

        <!-- End Position -->
        <text class="end" [attr.transform]="translate(width - 2, 0)">
          {{ reference.commonName }}, Chr{{ chromosome }}:{{ interval.end }} b
        </text>

        <!-- Vertical line extending end point of reference interval -->
        <path [attr.d]="getVLinePath(width, trackHeight + 43, -8)"></path>
      </g>

      <text class="all-caps" [attr.transform]="translate(2, 2*trackHeight + 54)">{{ comparison.commonName }}</text>

      <!-- Interval coordinates for current view extents in comparison positions -->
      <g *ngIf="interval" id="comparison-interval-coordinates" [attr.transform]="translate(0, 2*trackHeight + 65)">
        <!-- Vertical line extending start point of comparison interval -->
        <path [attr.d]="getVLinePath(0, -(trackHeight + 36), 1)"></path>

        <!-- Start position -->
        <text [attr.transform]="translate(2, 0)">
          {{ comparison.commonName }}, Chr{{ interval.compStart.chr }}:{{ interval.compStart.loc }} b
        </text>

        <!-- End Position -->
        <text class="end" [attr.transform]="translate(width - 2, 0)">
          {{ comparison.commonName }}, Chr{{ interval.compEnd.chr }}:{{ interval.compEnd.loc }} b
        </text>

        <!-- Vertical line extending end point of comparison interval -->
        <path [attr.d]="getVLinePath(width, -(trackHeight + 36), 1)"></path>
      </g>
    </g>

    <!-- Comparison genome legend -->
    <g id="comparison-genome-legend" [attr.transform]="translate(0, legendOffset)">
      <!-- Title -->
      <text [attr.transform]="translate(width / 2, 0)"> {{ comparison.commonName }} Genome Key </text>

      <!-- Legend -->
      <g id="legend" [attr.transform]="translate(getLegendXTrans(), 20)">
        <g *ngFor="let chr of getCompChromosomes()" [attr.transform]="translate(getLegendItemXTrans(chr), 0)"
           (mouseenter)="(isChrActive(chr)) ? currentChromosome = chr : null" (mouseleave)="currentChromosome = null">
          <!-- Colored circles -->
          <circle r="10" [attr.fill]="genomeColors[chr]" [attr.fill-opacity]="getLegendItemOpacity(chr)"></circle>

          <!-- Chromosome label -->
          <text [attr.transform]="translate(0, 25)" [attr.fill-opacity]="getLegendItemOpacity(chr)"> {{ chr }} </text>
        </g>
      </g>
    </g>

    <!-- Progress bar and loading overlay -->
    <g *ngIf="!(blocks && referenceGenes && comparisonGenes)" id="loading-overlay">
      <rect [attr.width]="width" [attr.height]="height"></rect>

      <g id="progress-bar" [attr.transform]="translate(width * 0.05, height / 2)">
        <rect [attr.width]="width * 0.9" height="5" rx="2" ry="2" fill-opacity="0.4"></rect>
        <rect [attr.width]="progress * (width * 0.9)" height="5" rx="2" ry="2" fill-opacity="0.7"></rect>
      </g>
    </g>
  </svg>
</div>
